File: CinemaTicket/src/domain/models/priceTable.ts
import { VisitorType } from './visitor';

type PriceList = {[key: string]: number}

const MovieDayPriceTable: PriceList = {
  [VisitorType.general]: 1800,
  [VisitorType.member]: 1100,
  [VisitorType.senior]: 1100,
  [VisitorType.seniorMember]: 1000,
  [VisitorType.disabledPerson]: 1000,
  [VisitorType.college]: 1500,
  [VisitorType.highSchoolStudent]: 1000,
  [VisitorType.juniorHighSchoolStudent]: 1000,
  [VisitorType.elementaryStudent]: 1000,
};
const WorkdayCommon: PriceList = {
  [VisitorType.general]: 1800,
  [VisitorType.member]: 1000,
  [VisitorType.senior]: 1100,
  [VisitorType.seniorMember]: 1000,
  [VisitorType.disabledPerson]: 1000,
  [VisitorType.college]: 1500,
  [VisitorType.highSchoolStudent]: 1000,
  [VisitorType.juniorHighSchoolStudent]: 1000,
  [VisitorType.elementaryStudent]: 1000,
};
const WorkdayLateShow: PriceList = {
  [VisitorType.general]: 1800,
  [VisitorType.member]: 1000,
  [VisitorType.senior]: 1100,
  [VisitorType.seniorMember]: 1000,
  [VisitorType.disabledPerson]: 1000,
  [VisitorType.college]: 1500,
  [VisitorType.highSchoolStudent]: 1000,
  [VisitorType.juniorHighSchoolStudent]: 1000,
  [VisitorType.elementaryStudent]: 1000,
};
const HolidayCommon: PriceList = {
  [VisitorType.general]: 1800,
  [VisitorType.member]: 1300,
  [VisitorType.senior]: 1100,
  [VisitorType.seniorMember]: 1000,
  [VisitorType.disabledPerson]: 1000,
  [VisitorType.college]: 1500,
  [VisitorType.highSchoolStudent]: 1000,
  [VisitorType.juniorHighSchoolStudent]: 1000,
  [VisitorType.elementaryStudent]: 1000,
};
const HolidayLateShow: PriceList = {
  [VisitorType.general]: 1800,
  [VisitorType.member]: 1000,
  [VisitorType.senior]: 1100,
  [VisitorType.seniorMember]: 1000,
  [VisitorType.disabledPerson]: 1000,
  [VisitorType.college]: 1500,
  [VisitorType.highSchoolStudent]: 1000,
  [VisitorType.juniorHighSchoolStudent]: 1000,
  [VisitorType.elementaryStudent]: 1000,
};

const MovieDay = 1;

const SaturDay = 6;
const SunDay = 0;

const LateShowStartAt = 22;

export default class PriceTable {
  // XXX: screeningTimeの命名が糞
  static get(screeningTime: Date): PriceList {
    if (screeningTime.getDate() === MovieDay) {
      return MovieDayPriceTable;
    }

    // TODO: 祝日は後で考える
    if (screeningTime.getDay() === SunDay || screeningTime.getDay() === SaturDay) {
      return screeningTime.getHours() < LateShowStartAt ? HolidayCommon : HolidayLateShow;
    }

    return screeningTime.getHours() < LateShowStartAt ? WorkdayCommon : WorkdayLateShow;
  }
}

File: CinemaTicket.kt/src/domain/models/priceTable.kt
package domain.models

import java.time.DayOfWeek
import java.time.LocalDateTime


class PriceTable() {
    companion object {
        const val MOVIE_DAY = 1

        val MovieDayPriceTable = PriceList(
            mapOf<VisitorType, Int>(
                VisitorType.General to 1800,
                VisitorType.Member to 1100,
                VisitorType.Senior to 1100,
                VisitorType.SeniorMember to 1000,
                VisitorType.DisabledPerson to 1000,
                VisitorType.College to 1500,
                VisitorType.HighSchoolStudent to 1000,
                VisitorType.JuniorHighSchoolStudent to 1000,
                VisitorType.ElementaryStudent to 1000
            )
        )

        val WorkdayCommon = PriceList(
            mapOf<VisitorType, Int>(
                VisitorType.General to 1800,
                VisitorType.Member to 1000,
                VisitorType.Senior to 1100,
                VisitorType.SeniorMember to 1000,
                VisitorType.DisabledPerson to 1000,
                VisitorType.College to 1500,
                VisitorType.HighSchoolStudent to 1000,
                VisitorType.JuniorHighSchoolStudent to 1000,
                VisitorType.ElementaryStudent to 1000
            )
        )

        val WorkdayLateShow = PriceList(
            mapOf<VisitorType, Int>(
                VisitorType.General to 1800,
                VisitorType.Member to 1000,
                VisitorType.Senior to 1100,
                VisitorType.SeniorMember to 1000,
                VisitorType.DisabledPerson to 1000,
                VisitorType.College to 1500,
                VisitorType.HighSchoolStudent to 1000,
                VisitorType.JuniorHighSchoolStudent to 1000,
                VisitorType.ElementaryStudent to 1000
            )
        )

        val HolidayCommon = PriceList(
            mapOf<VisitorType, Int>(
                VisitorType.General to 1800,
                VisitorType.Member to 1300,
                VisitorType.Senior to 1100,
                VisitorType.SeniorMember to 1000,
                VisitorType.DisabledPerson to 1000,
                VisitorType.College to 1500,
                VisitorType.HighSchoolStudent to 1000,
                VisitorType.JuniorHighSchoolStudent to 1000,
                VisitorType.ElementaryStudent to 1000
            )
        )

        val HolidayLateShow = PriceList(
            mapOf<VisitorType, Int>(
                VisitorType.General to 1800,
                VisitorType.Member to 1000,
                VisitorType.Senior to 1100,
                VisitorType.SeniorMember to 1000,
                VisitorType.DisabledPerson to 1000,
                VisitorType.College to 1500,
                VisitorType.HighSchoolStudent to 1000,
                VisitorType.JuniorHighSchoolStudent to 1000,
                VisitorType.ElementaryStudent to 1000
            )
        )
    }

    fun get(screeningTime: LocalDateTime): PriceList {
        if (screeningTime.dayOfMonth == MOVIE_DAY) {
            return MovieDayPriceTable
        }

        if (screeningTime.dayOfWeek == DayOfWeek.SUNDAY || screeningTime.dayOfWeek == DayOfWeek.SATURDAY) {
           return if (screeningTime.hour < 22) HolidayCommon else HolidayLateShow
        }
        return if (screeningTime.hour < 22) WorkdayCommon else WorkdayLateShow
    }
}

class PriceList(val priceList: Map<VisitorType, Int>)
